lib/TH/generic/THTensorMath.c

```diff
@@ -1192,12 +1277,24 @@ void THTensor_(sum)(THTensor *r_, THTensor *t, int dimension)
   THTensor_(resize)(r_, dim, NULL);
   THLongStorage_free(dim);
 
-  TH_TENSOR_DIM_APPLY2(real, t, real, r_, dimension,
-                       accreal sum = 0;
-                       long i;
-                       for(i = 0; i < t_size; i++)
-                         sum += t_data[i*t_stride];
-                       *r__data = (real)sum;);
+  // two implementations optimized for data locality
+  if (t->stride[dimension] == 1) {
+    TH_TENSOR_DIM_APPLY2(real, t, real, r_, dimension,
+                         accreal sum = 0;
+                         long i;
+                         for(i = 0; i < t_size; i++)
+                           sum += t_data[i*t_stride];
+                         *r__data = (real)sum;);
+  } else {
+    THTensor_(zero)(r_);
+    THTensor *temp_ = THTensor_(newWithTensor)(r_);
+    // r_.expand_as(t)
+    temp_->size[dimension] = t->size[dimension];
+    temp_->stride[dimension] = 0;
+
+    THTensor_(cadd)(temp_, temp_, 1, t);
```

Comment:
Good question. I know it's currently correct, but the question is whether it relies on the implementation. It comes down to whether `cadd` behavior is defined for destinations with stride-0 dimensions. Currently it does (at least for the CPU), because it only parallelizes the contiguous case.

Perhaps it would be cleaner to use `TH_APPLY` here directly instead of `cadd`, because we can require that `TH_APPLY` handles stride-0 destination properly. If we want to parallelize it in the future we'll need to split it into two versions.